### PR TITLE
Added fix in preact/hooks. Failing on preact-render-to-string when us…

### PR DESCRIPTION
**Pull request by gunins**
When use  preact-render-to-string in service workers as `iife` Getting errors cannot set property __hooks of undefined.

### CODE REVIEW

**Comment by JoviDeCroock** [comment]
There are a lot of unrelated changes in here

**Comment by gunins** [comment]
Appologies, auto formating.

**Comment by JoviDeCroock** [comment]
I'm not entirely sure how this is possible, what version of `preact-render-to-string` are you using? And what version of `Preact`?

**Comment by gunins** [comment]
Latest

**Comment by gunins** [comment]
Me too, was very surprised, nodejs with require was fine, but in service worker rollup, got error.

**Comment by JoviDeCroock** [comment]
We'll have to see a reproduction since this might not be the best place to fix this

**Comment by gunins** [comment]
Agree, this fix was working for me, but not sure if the best way to fix. But I really appreciate, if `preact-render-to-string` will work in browser too. Like I said working in nodejs environment.

**Comment by JoviDeCroock** [comment]
It does work in browser, this isn't a fix relating to making it work in a browser. This is an underlying or usage issue I think.

**Comment by developit** [comment]
Hiya. If your finding a situation where hooks don't have a value for `currentComponent`, there's no change in Preact that can fix that. This happens when a hook function is called by a function that isn't in the render path.

In the case you described, what's happening is that you're using `preact/hooks` in a way where it can't see the copy of Preact you're using with render-to-string.

Can you post a sample of how you are loading preact, preact/hooks and preact-render-to-string?

**Comment by gunins** [comment]
That’s was in my mind. I try to create basic example, and publish in github. Or option two, try to figure out reason for error, and publish too.

**Comment by gunins** [comment]
Hi! I figured out to create same issue in example. https://github.com/gunins/preact-test this one is compiled file https://github.com/gunins/preact-test/blob/master/dist/index.js
Just run this cile in browser, and check `console.log`
Also I have to admit, this could be edge case, because I solve by using `rollup-plugin-node-resolve` instead of `rollup-plugin-includepaths`,  but there is some issue, why direct include treating differently than compiled.

**Comment by JoviDeCroock** [comment]
Why are you not calling createElement on App in your render initiation?

**Comment by gunins** [comment]
Basically, I did different tests, in examples, and readme not mentioned, createElement is required. But added createElement https://github.com/gunins/preact-test/blob/master/src/index.ts, same error.

File: hooks/src/index.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ options._render = vnode => {
 	currentComponent = vnode._component;
 	currentIndex = 0;
 
-	if (currentComponent.__hooks) {
+	if (currentComponent !== undefined && currentComponent.__hooks) {
```
Comment (contedenis) [inline hooks/src/index.js:29]: You don't need to compare with undefined.
```
if (currentComponent && currentComponent.__hooks)
```

File: hooks/src/index.js [hunk 2]
```diff
@@ -129,7 +130,7 @@ export function useState(initialState) {
 export function useReducer(reducer, initialState, init) {
 	/** @type {import('./internal').ReducerHookState} */
 	const hookState = getHookState(currentIndex++, 2);
-	if (!hookState._component) {
+	if (hookState !== undefined && !hookState._component) {
```
Comment (contedenis) [inline hooks/src/index.js:133]: same

Comment (gunins) [reply hooks/src/index.js]: Changed.

File: hooks/src/index.js (hunk 1 above)
Comment (gunins) [reply hooks/src/index.js]: Changed

**Comment by JoviDeCroock** [comment]
@contedenis I found the issue you are using our source from preact instead of the build output. This meant that the options hook wasn't getting called replacing

```
const include = {
	'preact-render-to-string': './node_modules/preact-render-to-string',
	'preact':       './node_modules/preact',
	'preact/hooks': './node_modules/preact/hooks'
}
```

Fixes it

**Comment by contedenis** [comment]
> @contedenis I found the issue you are using our source from preact instead of the build output. This meant that the options hook wasn't getting called replacing
> 
> ```
> const include = {
> 	'preact-render-to-string': './node_modules/preact-render-to-string',
> 	'preact':       './node_modules/preact',
> 	'preact/hooks': './node_modules/preact/hooks'
> }
> ```
> 
> Fixes it

The issue was reported by @gunins, I just commented on his PR.

**Comment by gunins** [comment]
> > @contedenis I found the issue you are using our source from preact instead of the build output. This meant that the options hook wasn't getting called replacing
> > ```
> > const include = {
> > 	'preact-render-to-string': './node_modules/preact-render-to-string',
> > 	'preact':       './node_modules/preact',
> > 	'preact/hooks': './node_modules/preact/hooks'
> > }
> > ```
> > 
> > 
> > Fixes it
> 
> The issue was reported by @gunins, I just commented on his PR.
Actually this fix still breaking build, but I get the point, building by using source will fail.
Thanks!

**Comment by JoviDeCroock** [comment]
@gunins can we close this PR or are you still seeing issues?

**Comment by gunins** [comment]
Yes you can close. Will be helpful, if in `preact-render-to-string` documentation put note, doesn't work from source, need to use compiled version. Thanks again!

**Comment by JoviDeCroock** [comment]
@gunins see linked PR https://github.com/preactjs/preact/pull/2497
